engines/cryomni3d/versailles/engine.cpp

```diff
@@ -0,0 +1,1443 @@
+	}
+
+	const Graphics::Surface *image = imageDecoder->getSurface();
+
+	// Duplicate image to let hook modify it
+	Graphics::ManagedSurface dstSurface(image->w, image->h, image->format);
+	dstSurface.blitFrom(*image);
+
+	delete imageDecoder;
+	imageDecoder = nullptr;
+
+	if (hook) {
+		(this->*hook)(dstSurface);
+	}
+
+	g_system->copyRectToScreen(dstSurface.getPixels(), dstSurface.pitch, 0, 0,
+	                           dstSurface.w, dstSurface.h);
+	g_system->updateScreen();
+
+	setMousePos(Common::Point(320, 240)); // Center of screen
+	setCursor(181);
+
+	g_system->showMouse(true);
+
+	bool exitImg = false;
+	while (!g_engine->shouldQuit() && !exitImg) {
+		if (pollEvents()) {
+			if (getCurrentMouseButton() == 1) {
+				exitImg = true;
+			}
+		}
+		g_system->updateScreen();
+	}
+	waitMouseRelease();
+	clearKeys();
+
+	g_system->showMouse(false);
+	setMousePos(Common::Point(320, 240)); // Center of screen
+}
+
+void CryOmni3DEngine_Versailles::executeSeeAction(unsigned int actionId) {
+	if (_currentLevel == 7 && _currentPlaceId != 20) {
+		// Not enough time for paintings
+		displayMessageBoxWarp(14);
+		return;
+	}
+
+	const FixedImgCallback &cb = _imgScripts.getVal(actionId, nullptr);
+	if (cb != nullptr) {
+		handleFixedImg(cb);
+	} else {
+		warning("Image script %u not found", actionId);
+	}
+}
+
+void CryOmni3DEngine_Versailles::executeSpeakAction(unsigned int actionId) {
+	PlaceActionKey key(_currentPlaceId, actionId);
+	Common::HashMap<PlaceActionKey, Common::String>::iterator it = _whoSpeaksWhere.find(key);
+	g_system->showMouse(true);
+	bool doneSth = false;
+	if (it != _whoSpeaksWhere.end()) {
+		doneSth = _dialogsMan.play(it->_value);
+	}
+	g_system->showMouse(false);
+	_forcePaletteUpdate = true;
+	if (doneSth) {
+		setMousePos(Common::Point(320, 240)); // Center of screen
+	}
+}
+
+void CryOmni3DEngine_Versailles::executeDocAction(unsigned int actionId) {
+	if (_currentLevel == 7) {
+		// Not enough time for doc
+		displayMessageBoxWarp(13);
+		return;
+	}
+
+	Common::HashMap<unsigned int, const char *>::iterator it = _docPeopleRecord.find(actionId);
+	if (it == _docPeopleRecord.end() || !it->_value) {
+		warning("Missing documentation record for action %u", actionId);
+		return;
+	}
+
+	_docManager.handleDocInGame(it->_value);
+
+	_forcePaletteUpdate = true;
+	setMousePos(Common::Point(320, 240)); // Center of screen
+}
+
+void CryOmni3DEngine_Versailles::handleFixedImg(const FixedImgCallback &callback) {
+	if (!callback) {
+		return;
+	}
+
+	ZonFixedImage::CallbackFunctor *functor =
+	    new Common::Functor1Mem<ZonFixedImage *, void, CryOmni3DEngine_Versailles>(this, callback);
+	_fixedImage->run(functor);
+	// functor is deleted in ZoneFixedImage
+	functor = nullptr;
+
+	if (_nextPlaceId == -1u) {
+		_forcePaletteUpdate = true;
+	}
+}
+
+unsigned int CryOmni3DEngine_Versailles::getFakeTransition(unsigned int actionId) const {
+	for (const FakeTransitionActionPlace *ft = kFakeTransitions; ft->actionId != nullptr; ft++) {
```

Comment:
`ft->actionId != nullptr` results in a compilation failure when building with GCC's `-std=c++11`. `ft->actionId` is not a pointer.